### PR TITLE
Faster broadcast preprocess

### PR DESCRIPTION
**Pull request by Thenerdstation**
**Context:**
Previously, the `_preprocess` would take upwards of 20% of the tape construction time because it was always building all of the wire patterns were always built with each function call. We remove this overhead, resulting in 15% speedup in tape construction, and ~2-3% e2e speedups for all interfaces.

Results from my laptop:

```
-      13.1±0.9ms       11.8±0.1ms     0.90  asv.core_suite.CircuitEvaluation.time_circuit(10, 3)
       23.6±0.3ms       21.8±0.5ms     0.93  asv.core_suite.CircuitEvaluation.time_circuit(10, 6)
       34.4±0.2ms       31.4±0.1ms     0.91  asv.core_suite.CircuitEvaluation.time_circuit(10, 9)
       3.68±0.2ms       3.48±0.2ms     0.95  asv.core_suite.CircuitEvaluation.time_circuit(2, 3)
       5.58±0.3ms       5.29±0.2ms     0.95  asv.core_suite.CircuitEvaluation.time_circuit(2, 6)
       7.46±0.3ms       7.24±0.3ms     0.97  asv.core_suite.CircuitEvaluation.time_circuit(2, 9)
       7.08±0.2ms       6.75±0.4ms     0.95  asv.core_suite.CircuitEvaluation.time_circuit(5, 3)
-      12.9±0.5ms      11.1±0.02ms     0.86  asv.core_suite.CircuitEvaluation.time_circuit(5, 6)
       17.2±0.3ms       16.0±0.3ms     0.93  asv.core_suite.CircuitEvaluation.time_circuit(5, 9)
       5.12±0.3ms       4.93±0.2ms     0.96  asv.core_suite.GradientComputation.time_gradient_autograd(2, 3)
       7.90±0.3ms       7.54±0.3ms     0.96  asv.core_suite.GradientComputation.time_gradient_autograd(2, 6)
       11.0±0.4ms       10.2±0.3ms     0.93  asv.core_suite.GradientComputation.time_gradient_autograd(5, 3)
       19.6±0.8ms         18.8±1ms     0.96  asv.core_suite.GradientComputation.time_gradient_autograd(5, 6)
       16.2±0.1ms       16.3±0.1ms     1.01  asv.core_suite.GradientComputation.time_gradient_tf(2, 3)
       28.1±0.4ms       28.8±0.6ms     1.03  asv.core_suite.GradientComputation.time_gradient_tf(2, 6)
       44.7±0.2ms       44.6±0.3ms     1.00  asv.core_suite.GradientComputation.time_gradient_tf(5, 3)
       88.1±0.5ms       87.3±0.4ms     0.99  asv.core_suite.GradientComputation.time_gradient_tf(5, 6)
       7.39±0.3ms       7.45±0.3ms     1.01  asv.core_suite.GradientComputation.time_gradient_torch(2, 3)
       19.5±0.6ms       18.6±0.9ms     0.95  asv.core_suite.GradientComputation.time_gradient_torch(2, 6)
       36.4±0.3ms       35.9±0.1ms     0.99  asv.core_suite.GradientComputation.time_gradient_torch(5, 3)
          129±2ms          126±2ms     0.98  asv.core_suite.GradientComputation.time_gradient_torch(5, 6)
```

**Description of the Change:**
Moved the wire pattern constructions to lambda methods, removing the overhead from all preprocess calls.

**Benefits:**
It's faster.

**Possible Drawbacks:**
None

**Related GitHub Issues:**
None

### CODE REVIEW

File: pennylane/templates/broadcast.py [hunk 1]
```diff
@@ -25,7 +25,7 @@
 from pennylane.templates.utils import get_shape
 from pennylane.wires import Wires
 
-OPTIONS = ["single", "double", "double_odd", "chain", "ring", "pyramid", "all_to_all", "custom"]
+OPTIONS = {"single", "double", "double_odd", "chain", "ring", "pyramid", "all_to_all", "custom"}
```
Comment (josh146) [inline pennylane/templates/broadcast.py:28]: is this just for faster 'contains' checking?

File: pennylane/templates/broadcast.py [hunk 2]
```diff
@@ -82,41 +105,16 @@ def _preprocess(parameters, pattern, wires):
         wire_sequence, parameters: preprocessed pattern and parameters
     """
 
-    custom_pattern = None
-
     if isinstance(pattern, str):
+        _wires = wires
```
Comment (josh146) [inline pennylane/templates/broadcast.py:109]: What's the reason for making this 'private'?

**Review by josh146 — APPROVED** [review]
Such a big win from a small change!

**Review by mariaschuld — APPROVED** [review]
Nice one!

File: pennylane/templates/broadcast.py [hunk 3]
```diff
@@ -82,41 +105,16 @@ def _preprocess(parameters, pattern, wires):
...
         if pattern not in OPTIONS:
             raise ValueError(f"did not recognize pattern {pattern}".format())
     else:
         # turn custom pattern into list of Wires objects
-        custom_pattern = [Wires(w) for w in pattern]
+        _wires = [Wires(w) for w in pattern]
```
Comment (mariaschuld) [inline pennylane/templates/broadcast.py:114]: Actually, in the new approach of being stingy with Wires parsing we could just remove this line altogether. This was in some sense just a check if the pattern has unique wires in every set. But if not, the gate will later complain, so this should be fine!

File: pennylane/templates/broadcast.py (hunk 1 above)
Comment (Thenerdstation) [reply pennylane/templates/broadcast.py:28]: Yes. By making it a set makes the `if x in OPTIONS` time O(1)

File: pennylane/templates/broadcast.py (hunk 2 above)
Comment (Thenerdstation) [reply pennylane/templates/broadcast.py:109]: I couldn't come up with a better name lol. I'm open to suggestions.

File: pennylane/templates/broadcast.py (hunk 3 above)
Comment (Thenerdstation) [reply pennylane/templates/broadcast.py:114]: so instead just set `_wires = pattern`?

**Pull request merged**